pkg/provider/azure_controller_common.go

```diff
@@ -267,7 +279,38 @@ func (c *controllerCommon) AttachDisk(ctx context.Context, async bool, diskName,
 			klog.Warningf("azureDisk - switch to batch operation due to rate limited, QPS: %f", c.diskOpRateLimiter.QPS())
 		}
 	}
-	return lun, vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+
+	backoffManager := wait.Backoff{
+		Duration: updateVMUpdateRetryDuration,
+		Factor:   updateVMUpdateRetryFactor,
+		Steps:    updateVMUpdateRetrySteps,
+	}
+
+	err = vmset.WaitForUpdateResult(ctx, future, nodeName, "attach_disk")
+	for c.vmUpdateRequired(future, err) {
+		select {
+		case <-ctx.Done():
+			return -1, context.DeadlineExceeded
+		case <-time.After(backoffManager.Step()):
```

Comment:
Even if backoff ends, the plan was to loop until context timeout if updates continue failing due to `OperationPreempted` error.  We back off with a pretty aggressive factor here, so by the time backoff exhausts all its steps, it would have a `3^5`   seconds of wait time. As we discussed previously, there is no guarantee of when the operation is going to succeed, so it was hard to put a hard cap on how many update retries we should attempt. Then, we believed it would be reasonable to retry either until success or the given context times out.

Please let me know what you think.

Reply:
thanks for the explanation. I think if backoff ends, it should exit, if first few tries failed, the next try will also fail probably per my past experience, you may refer to this, and use defaultBackoff if that's feasible.

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/8fb8dd8d3f23872587607436d02f938f3c6e9725/pkg/provider/azure_managedDiskController.go#L257-L273

Reply:
I kept the backoff configurations for attach as is because I felt that factor of `1.5` was too small for handling `OperationPreempted` errors and thought we would hammering the front end. But I updated the loop to use `ExponentialBackoffWithContext`.